test/test.ts

```diff
@@ -0,0 +1,119 @@
+let reference: ReferenceTranscoder;
+
+async function setup<T>(wasmPath: string, width = WIDTH, height = HEIGHT): Promise<T> {
+  numBlocks = ((width + 3) >> 2) * ((height + 3) >> 2);
+
+  const compressedByteLength = numBlocks * 16;
+  const uncompressedByteLength = WIDTH * ((HEIGHT + 3) >> 2) * 4 * 4;
+  const totalByteLength = compressedByteLength + uncompressedByteLength;
+  const memory = new WebAssembly.Memory({ initial: ((totalByteLength + 65535) >> 16) + 1 });
+
+  compressedView = new Uint8Array(memory.buffer, 65536, compressedByteLength);
+  uncompressedView = new Uint8Array(memory.buffer, 65536 + compressedByteLength, uncompressedByteLength);
+
+  // Generator.
+  const generatorModule = new WebAssembly.Module(fs.readFileSync('build/uastc_generator.wasm'));
+  generator = new WebAssembly.Instance(generatorModule, { env: { memory: memory } }).exports as unknown as IGenerator;
+
+  // Reference transcoder.
+  reference = new ReferenceTranscoder(WIDTH, HEIGHT);
+  await reference.init();
+
+  // Source transcoder.
+  const module = new WebAssembly.Module(fs.readFileSync(wasmPath));
+  return new WebAssembly.Instance(module, {env: { memory: memory }}).exports as unknown as T;
+}
+
+test('ASTC4x4', async (t) => {
+  const transcoder = await setup<ITranscoder>('build/uastc_astc.wasm');
+
+  for (let m = 0; m < NUM_MODES; m++) {
+    generator.generate(m, numBlocks);
+    const expected = reference.transcode(compressedView.slice(), reference.TranscodeTarget.ASTC_4x4_RGBA);
+    t.equals(transcoder.transcode(numBlocks), 0, `mode ${m}: ok`);
+    t.ok(arrayEquals(compressedView, expected), `mode ${m}: data`);
+  }
+
+  t.end();
+});
+
+test('BC7', async (t) => {
+  const transcoder = await setup<ITranscoder>('build/uastc_bc7.wasm');
+
+  for (let m = 0; m < NUM_MODES; m++) {
+    generator.generate(m, numBlocks);
+    const expected = reference.transcode(compressedView.slice(), reference.TranscodeTarget.BC7_RGBA);
+    t.equals(transcoder.transcode(numBlocks), 0, `mode ${m}: ok`);
+    t.ok(arrayEquals(compressedView, expected), `mode ${m}: data`);
+  }
+
+  t.end();
+});
+
+test('RGBA32', async (t) => {
+  const decoder = await setup<IDecoder>('build/uastc_rgba32_unorm.wasm');
+
+  for (let m = 0; m < NUM_MODES; m++) {
+    generator.generate(m, numBlocks);
+    const expected = reference.transcode(compressedView.slice(), reference.TranscodeTarget.RGBA32);
+    t.equals(decoder.decodeRGBA32(WIDTH, HEIGHT), 0, `mode ${m}: ok`);
+    t.ok(arrayEquals(uncompressedView, expected), `mode ${m}: data`);
+  }
+
+  t.end();
+});
+
+test('RGBA32 - npot', async (t) => {
+  const decoder = await setup<IDecoder>('build/uastc_rgba32_unorm.wasm', WIDTH - 1, HEIGHT - 1);
+
+  for (let m = 0; m < NUM_MODES; m++) {
+    generator.generate(m, numBlocks);
+    const expected = reference.transcode(compressedView.slice(), reference.TranscodeTarget.RGBA32);
+    t.equals(decoder.decodeRGBA32(WIDTH, HEIGHT), 0, `mode ${m}: ok`); // TODO: Why DIM, not DIM - 1?
```

Comment:
I would've expected this to require `WIDTH - 1`  and `HEIGHT - 1`, but it passes as-is, and not with the change. Any idea why?

Reply:
Hang on, that's a typo...